source/base/polynomials_rt_bubbles.cc

```diff
@@ -0,0 +1,504 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+#include "deal.II/base/polynomials_rt_bubbles.h"
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/thread_management.h>
+#include <iostream>
+#include <iomanip>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+template <int dim>
+PolynomialsRT_Bubbles<dim>::PolynomialsRT_Bubbles (const unsigned int k)
+  :
+  my_degree(k),
+  raviart_thomas_space (k-1),
+  monomials(k+2),
+  n_pols(compute_n_pols(k))
+{
+  Assert (dim >= 2, ExcImpossibleInDim(dim));
+
+  for (unsigned int i=0; i<monomials.size(); ++i)
+    monomials[i] = Polynomials::Monomial<double> (i);
+}
+
+
+template <int dim>
+void
+PolynomialsRT_Bubbles<dim>::compute (const Point<dim>         &unit_point,
+                                     std::vector<Tensor<1,dim> > &values,
+                                     std::vector<Tensor<2,dim> > &grads,
+                                     std::vector<Tensor<3,dim> > &grad_grads,
+                                     std::vector<Tensor<4,dim> > &third_derivatives,
+                                     std::vector<Tensor<5,dim> > &fourth_derivatives) const
+{
+  Assert(values.size()==n_pols || values.size()==0,
+         ExcDimensionMismatch(values.size(), n_pols));
+  Assert(grads.size()==n_pols || grads.size()==0,
+         ExcDimensionMismatch(grads.size(), n_pols));
+  Assert(grad_grads.size()==n_pols || grad_grads.size()==0,
+         ExcDimensionMismatch(grad_grads.size(), n_pols));
+  Assert(third_derivatives.size()==n_pols || third_derivatives.size()==0,
+         ExcDimensionMismatch(third_derivatives.size(), n_pols));
+  Assert(fourth_derivatives.size()==n_pols || fourth_derivatives.size()==0,
+         ExcDimensionMismatch(fourth_derivatives.size(), n_pols));
+
+  // Third and fourth derivatives are not implemented
+  (void)third_derivatives;
+  Assert(third_derivatives.size()==0,
+         ExcNotImplemented());
+  (void)fourth_derivatives;
+  Assert(fourth_derivatives.size()==0,
+         ExcNotImplemented());
+
+  const unsigned int n_sub = raviart_thomas_space.n();
+
+  // Guard access to the scratch arrays in the following block
```

Comment:
You don't use the `p_*` arrays all the way to the end of the function. Consequently, it may be useful to release the mutex as soon as you can, and for this, just place a `{` here...